### ISSUE TITLE
Combination of `PublishSingleFile` and `PublishReadyToRun` prevents app from starting

### ISSUE THREAD

**Issue by Abrynos** [dotnet/wpf#7282]
## Problem description

Since changing the `TargetFramework` in our .csproj file from `net6.0-windows` to `net7.0-windows` our app crashes upon starting with a `System.TypeInitializationException` thrown in the initializer for `System.Windows.Media.FontFamily`.

Removing either the `PublishSingleFile` or the `PublishReadyToRun` tag from the projects .csproj file (or switching back to `net6.0-windows`) fixes the problem.

## Stack-trace

```
System.TypeInitializationException: The type initializer for 'System.Windows.Media.FontFamily' threw an exception.
 ---> System.TypeInitializationException: The type initializer for 'MS.Internal.FontCache.DWriteFactory' threw an exception.
 ---> System.InvalidCastException: Specified cast is not valid.
   at MS.Internal.Text.TextInterface.Native.Util.ConvertHresultToException(Int32 hr)
   at MS.Internal.Text.TextInterface.Factory.Initialize(FactoryType factoryType)
   at MS.Internal.Text.TextInterface.Factory..ctor(FactoryType factoryType, IFontSourceCollectionFactory fontSourceCollectionFactory, IFontSourceFactory fontSourceFactory)
   at MS.Internal.FontCache.DWriteFactory..cctor()
   --- End of inner exception stack trace ---
   at MS.Internal.FontCache.DWriteFactory.get_SystemFontCollection()
   at System.Windows.Media.FontFamily..cctor()
   --- End of inner exception stack trace ---
   at MS.Internal.Text.DynamicPropertyReader.GetTypeface(DependencyObject element)
   at MS.Internal.Text.TextProperties.InitCommon(DependencyObject target)
   at MS.Internal.Text.TextProperties..ctor(FrameworkElement target, Boolean isTypographyDefaultValue)
   at System.Windows.Controls.TextBoxView.GetLineProperties()
   at System.Windows.Controls.TextBoxView.TextCache..ctor(TextBoxView owner)
   at System.Windows.Controls.TextBoxView.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at MS.Internal.Helper.MeasureElementWithSingleChild(UIElement element, Size constraint)
   at System.Windows.Controls.ScrollContentPresenter.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Grid.MeasureCell(Int32 cell, Boolean forceInfinityV)
   at System.Windows.Controls.Grid.MeasureCellsGroup(Int32 cellsHead, Size referenceSize, Boolean ignoreDesiredSizeU, Boolean forceInfinityV, Boolean& hasDesiredSizeUChanged)
   at System.Windows.Controls.Grid.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.ScrollViewer.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Border.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Control.MeasureOverride(Size constraint)
   at System.Windows.Controls.TextBox.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Grid.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at MS.Internal.Helper.MeasureElementWithSingleChild(UIElement element, Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Decorator.MeasureOverride(Size constraint)
   at System.Windows.Documents.AdornerDecorator.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Controls.Border.MeasureOverride(Size constraint)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Window.MeasureOverrideHelper(Size constraint)
   at System.Windows.Window.MeasureOverride(Size availableSize)
   at System.Windows.FrameworkElement.MeasureCore(Size availableSize)
   at System.Windows.UIElement.Measure(Size availableSize)
   at System.Windows.Interop.HwndSource.SetLayoutSize()
   at System.Windows.Interop.HwndSource.set_RootVisualInternal(Visual value)
   at System.Windows.Window.SetRootVisualAndUpdateSTC()
   at System.Windows.Window.SetupInitialState(Double requestedTop, Double requestedLeft, Double requestedWidth, Double requestedHeight)
   at System.Windows.Window.CreateSourceWindow(Boolean duringShow)
   at System.Windows.Window.ShowHelper(Object booleanBox)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at MS.Internal.CulturePreservingExecutionContext.Run(CulturePreservingExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
   at System.Windows.Application.RunDispatcher(Object ignore)
   at System.Windows.Application.RunInternal(Window window)
   at MyProject.App.Main()
```

## Expected behavior

The app starts the same way it did when using `net6.0-windows` as `TargetFramework`.

## net48

No idea whether this also reproduces in .NET Framework 4.8. We're using `net7.0` exclusively.

## Relation to Visual Studio (or other tooling)

None. We did our testing using the `dotnet publish` command.

## dotnet info:
```
.NET SDK:
 Version:   7.0.100
 Commit:    e12b7af219

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.100\
 
...

Environment variables:
  Not set

global.json file:
  Not found
```

<details>
  <summary>Minimal reproduction</summary>
  
There are no `Directory.Build.props` or `Directory.Packages.props` files present.

Executing the project results in a file named `ex.txt` with the stack-trace being left in the execution directory.

  Project.csproj

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net7.0-windows</TargetFramework>
  </PropertyGroup>

  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <UseWPF>true</UseWPF>
  </PropertyGroup>

  <PropertyGroup>
    <PublishSingleFile>true</PublishSingleFile>
    <PublishReadyToRun>true</PublishReadyToRun>
  </PropertyGroup>
</Project>
```

App.xaml

```xml
<Application x:Class="MyProject.App"
             xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             StartupUri="TestView.xaml"/>
```

App.xaml.cs

```csharp
using System;
using System.IO;
using System.Windows;

namespace MyProject;

public sealed partial class App
{
    protected override void OnStartup(StartupEventArgs e)
    {
        AppDomain.CurrentDomain.UnhandledException += OnUnhandledException;

        base.OnStartup(e);
    }

    private void OnUnhandledException(object sender, UnhandledExceptionEventArgs e)
    {
        ArgumentNullException.ThrowIfNull(e);
        ArgumentNullException.ThrowIfNull(e.ExceptionObject);

        Exception exception = (Exception)e.ExceptionObject;

        File.WriteAllText("ex.txt", exception.ToString());

        Environment.Exit(byte.MaxValue);
    }
}
```

TestView.xaml

```xml
<Window x:Class="MyProject.TestView"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        mc:Ignorable="d">
    <Grid>
        <TextBox Text="Test"/>
    </Grid>
</Window>
```

TestView.xaml.cs

```csharp
namespace MyProject;

public partial class TestView
{
    public TestView() => InitializeComponent();
}
```
</details>

special thanks to @beagle1986 for lending me his ear.

**Comment by lindexi** [comment]
@Abrynos The DirectWriteForwarder.dll is a C++\CLI project. Looks like it's not well supported.

**Comment by Gakki123** [comment]
I'm also experiencing this issue

**Comment by EgemenCiftci** [comment]
I think this is the same as the below issue.
[https://github.com/dotnet/runtime/issues/65879](https://github.com/dotnet/runtime/issues/65879)

**Comment by dragnilar** [comment]
We started experiencing this problem today when we tried publishing a UAT version of our application using Dot Net 7. Turning off R2R allows the application to publish without crashing on the MS.Internal.FontCache.DWriteFactory Initializer.

@EgemenCiftci  

Curious, were you able to publish with R2R turned on in Dot Net 6?

**Comment by dragnilar** [comment]
@EgemenCiftci I updated my previous comment since I wasn't very clear. You still managed to answer my question. Thanks! 😀

That is interesting that it did not work for you in Dot Net 6.  Does your app use any custom fonts?

**Comment by EgemenCiftci** [comment]
@dragnilar No. There are no fancy things. The bug can be reproduced with a default WPF app with a TextBlock on the MainWindow.

**Comment by FM1973** [comment]
I have the same problem. Using .net 7 the app crashes, using .net 6 the app starts as expected.
Also using R2R and single file.

**Comment by Mathyn** [comment]
Can confirm I have the same problem when TargetFramework is set to `net7.0-windows` and PublishReadyToRun is set to `true`. Any word on a fix?

**Comment by harikrishnan-p-v** [comment]
@Mathyn add this
 `<PropertyGroup>
    <IncludeNativeLibrariesForSelfExtract>true</IncludeNativeLibrariesForSelfExtract>
  </PropertyGroup>`

**Comment by ThomasGoulet73** [comment]
I'm closing this issue since the bug was fixed in dotnet/runtime for .Net 8 and the issue tracking this bug was closed (dotnet/runtime#65879). I am no longer able to reproduce this bug using the .Net 8 SDK (I get an Exception with 7.0.306 and no exception with  8.0.100-preview.7.23376.3).

The source of this bug was the fact that composite R2R had bug where it couldn't handle C++/CLI dlls in certain case and DirectWriteForwarder.dll, which is used by WPF, was one of those case.
